src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,265 @@
+
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/alter_partition_reassignments.h"
+
+#include "cluster/errc.h"
+#include "cluster/metadata_cache.h"
+#include "cluster/topics_frontend.h"
+#include "cluster/types.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_request.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_response.h"
+#include "kafka/server/errors.h"
+#include "kafka/server/fwd.h"
+#include "kafka/server/quota_manager.h"
+#include "model/metadata.h"
+#include "model/namespace.h"
+#include "model/timeout_clock.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sstring.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <chrono>
+#include <iterator>
+#include <system_error>
+#include <vector>
+
+namespace kafka {
+
+using partitions_request_iterator
+  = std::vector<reassignable_partition>::iterator;
+
+template<typename ResultIter, typename Predicate>
+partitions_request_iterator validate_replicas(
+  partitions_request_iterator begin,
+  partitions_request_iterator end,
+  ResultIter out_it,
+  error_code ec,
+  const ss::sstring& error_message,
+  Predicate&& p) {
+    auto valid_range_end = std::partition(
+      begin, end, std::forward<Predicate>(p));
+
+    std::transform(
+      valid_range_end,
+      end,
+      out_it,
+      [ec, &error_message](const reassignable_partition& partition) {
+          return reassignable_partition_response{
+            .partition_index = partition.partition_index,
+            .error_code = ec,
+            .error_message = error_message,
+          };
+      });
+    return valid_range_end;
+}
+
+template<typename ResultIter>
+void validate_topics(
+  std::vector<reassignable_topic>& reassignable_topics,
+  ResultIter resp_it,
+  std::vector<model::node_id> all_node_ids) {
+    for (reassignable_topic& topic : reassignable_topics) {
+        reassignable_topic_response topic_response{.name = topic.name};
+
+        // An undefined replicas vector is not an error, see "Replicas" in the
+        // AlterPartitionReassignmentsRequest schemata. So we put those elements
+        // first with std::partition.
+        auto valid_partitions_end = std::partition(
```

Comment:
Why do we need to reorder the input if theres no error?

Reply:
The `validate_replicas` method is run a non-null replicas vector. However, replicas can be null as per the schemata definition. So instead of checking for a null replicas vector in every call to `validate_replicas`, I decided to just put those elements first and use iterator magic so we run validation on defined replicas only.

Reply:
Makes sense to me